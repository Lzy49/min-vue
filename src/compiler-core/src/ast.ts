export const enum NODE_TYPE {
  INTERPOLATION = 'INTERPOLATION',
  SIMPLE_EXPRESSION = 'SIMPLE_EXPRESSION',
  ELEMENT = "ELEMENT",
  TEXT = "TEXT",
  ROOT = "ROOT",
  COMPOUND_EXPRESSION = "COMPOUND_EXPRESSION"
}