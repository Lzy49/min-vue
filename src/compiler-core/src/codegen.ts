import { NODE_TYPE } from "./ast";
import { isString } from '../../shared'
import { helperNameMap, TO_DISPLAY_STRING, CREATE_ELEMENT_BLOCK } from './runtimeHelpers'
export function generate(ast) {
  // 实现就是拆分
  const context = createCodegenContext();

  genFunctionPreamble(ast, context);
  context.push(`return `);
  const args = ['_ctx', '_cache']
  const argsSignature = args.join(', ')
  const functionName = 'render'
  context.push(`function ${functionName}(${argsSignature}){`)
  context.push(`return `)
  // 处理 ast -> h
  genNode(ast.codegenNode, context)
  context.push(`}`)
  return {
    code: context.code
  }
  function genFunctionPreamble(ast, context) {
    if (ast.helpers.length > 0) {
      const VueBinging = 'Vue'
      const helpersSignature = ast.helpers.map(helper => `${helperNameMap[helper]}: ${context.helper(helper)}`).join(',')
      context.push(`const { ${helpersSignature} } = ${VueBinging};`);
    }
  }
}
// 创建 上下文 
function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperNameMap[key]}`;
    },
  }
  return context
}
// 处理不同类型 Node
function genNode(node, context) {
  switch (node.type) {
    case NODE_TYPE.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NODE_TYPE.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break;
    case NODE_TYPE.ELEMENT:
      genElement(node, context);
      break;
    case NODE_TYPE.TEXT:
      genText(node, context);
      break;
    case NODE_TYPE.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
    default:
  }
}
function genElement(node, context) {
  const { tag, props, children } = node;

  context.push(context.helper(CREATE_ELEMENT_BLOCK))
  context.push('(')

  genNodeList(genNullableArgs([tag, props, children]), context);

  context.push(`)`);

}
function genInterpolation(node, context) {
  context.push(context.helper(TO_DISPLAY_STRING))
  context.push('(')
  genNode(node.content, context)
  context.push(')')
}
function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}
function genExpression(node, context) {
  context.push(node.content)
}

function genText(node, context) {
  context.push(`'${node.content}'`)
}

function genNodeList(nodes: any, context: any) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];

    if (isString(node)) {
      push(`${node}`);
    } else {
      genNode(node, context);
    }
    // node 和 node 之间需要加上 逗号(,)
    // 但是最后一个不需要 "div", [props], [children]
    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullableArgs(args) {
  let i = args.length - 1
  // 去掉多余的参数
  for (; i >= 0; i--) {
    // 说明是 null
    if (args[i]) {
      break
    }
  }
  // 将空参转换为null
  return args.slice(0, i + 1).map((arg) => arg || "null");
}