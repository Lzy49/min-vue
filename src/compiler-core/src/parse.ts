import { NODE_TYPE } from "./ast";
const delimiter = [
  "{{",
  "}}"
]
const enum TagType {
  START,
  END,
}
export const baseParse = (template: string) => {
  const context = createParserContext(template)
  let ancestor = [];
  return createRoot(paseChildren(context, ancestor))
}

function paseChildren(context, ancestor): any {
  const nodes = [];
  while (isEnd(context, ancestor)) {
    let node;
    const source = context.source
    if (source.startsWith(delimiter[0])) {
      node = paseInterpolation(context)
    } else if (source.startsWith('<')) {
      if (/^<[a-z]*/.test(source)) {
        node = paseElement(context, ancestor)
      }
    }
    if (!node) {
      node = paseText(context)
    }
    nodes.push(node)
  }
  return nodes
}
function isEnd(context, ancestor) {
  if (context.source.length <= 0) {
    return false;
  }
  for (let key = ancestor.length - 1; key >= 0; key--) {
    const item = ancestor[key]
    if (context.source.startsWith(`</${item}>`)) {
      if (key !== ancestor.length - 1) {
        throw new Error('span 没有闭合标签')
      }
      return false
    }
  }
  return true
}
function createParserContext(template: string): { source: string } {
  return {
    source: template as string
  }
}
function createRoot(children) {
  return {
    children,
    type: NODE_TYPE.ROOT,
    helpers: []
  }
}

function paseInterpolation(context) {
  const startDelimiter = delimiter[0]
  const endDelimiter = delimiter[1]
  // 先推到开始分隔符后
  advanceBy(context, startDelimiter.length)
  const endIndex = context.source.indexOf(endDelimiter);
  // 截取 content
  const rawContent = paseTextData(context, endIndex);
  const content = rawContent.trim();
  // 再推到结束分隔符后
  advanceBy(context, endDelimiter.length)
  // 返回
  return {
    type: NODE_TYPE.INTERPOLATION,
    content: {
      type: NODE_TYPE.SIMPLE_EXPRESSION,
      content
    }
  }
}

function advanceBy(context: any, index: number) {
  context.source = context.source.slice(index);
}

// 处理因为 缺少闭合报错
// 就是 确定在什么范围内，缺少
// 解决：收集每一层 element 的范围
function paseElement(context: any, ancestor): any {
  // context.source
  // 首先要解析出 tag
  let element: any = parseTag(context, TagType.START);
  ancestor.push(element.tag)
  // 处理子节点
  // element.tag
  element.children = paseChildren(context, ancestor)
  ancestor.pop(element.tag)
  parseTag(context, TagType.END);
  return element
}

function parseTag(context, type: TagType) {
  const result = /^<\/?([a-z]*)/i.exec(context.source)
  // 推推 source 
  advanceBy(context, result[0].length + 1)
  if (type === TagType.END) return;
  const tag = result[1]
  return {
    type: NODE_TYPE.ELEMENT,
    tag
  }
}

function paseText(context: any) {
  const source = context.source
  const endTags = [delimiter[0], '<']
  let endIndex = context.source.length;
  for (let item of endTags) {
    const index = source.indexOf(item);
    if (index !== -1 && source.indexOf(item) < endIndex) {
      endIndex = source.indexOf(item)
    }
  }
  // 1. 取出值
  const content = paseTextData(context, endIndex);
  // 2. 推进
  return {
    type: NODE_TYPE.TEXT,
    content
  }
}

function paseTextData(context: any, length) {
  const content = context.source.slice(0, length);
  // 获取结束
  advanceBy(context, length);
  return content;
}

