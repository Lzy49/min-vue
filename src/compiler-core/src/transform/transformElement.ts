import { NODE_TYPE } from "../ast";
import { CREATE_ELEMENT_BLOCK } from '../runtimeHelpers'

export function transformElement(node, context) {
  if (node.type === NODE_TYPE.ELEMENT) {
    return () => {
      const vnodeTag = `'${node.tag}'`;

      const vnodeProps = null;
      let vnodeChildren = null;
      if (node.children.length > 0) {
        if (node.children.length === 1) {
          const child = node.children[0];
          vnodeChildren = child;
        }
      }
      context.helper(CREATE_ELEMENT_BLOCK)
      node.codegenNode = {
        type: NODE_TYPE.ELEMENT,
        tag: vnodeTag,
        props: vnodeProps,
        children: vnodeChildren
      }
   
    };
  }
}