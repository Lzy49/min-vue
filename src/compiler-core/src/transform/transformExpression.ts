import { NODE_TYPE } from "../ast";

export function transformExpression(node) {
  if (node.type === NODE_TYPE.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`;

  return node
}