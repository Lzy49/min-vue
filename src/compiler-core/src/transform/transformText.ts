import { NODE_TYPE } from '../ast'
import { isText } from './shared'
export const transformText = (node, context) => {
  // 判断el类型，该类型的子节点需要处理
  if (node.type === NODE_TYPE.ELEMENT) {
    const children = node.children;
    return () => {
      for (let i = 0; i < children.length; i++) {
        const item = children[i]
        // 如果子节点是 文本类型则需要进行转换。
        if (isText(item)) {
          let currentContainer;
          for (let j = i + 1; j < children.length; j++) {
            // 如果相邻的节点不是一个 文本则不需要进行处理
            const next = children[j];
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NODE_TYPE.COMPOUND_EXPRESSION,
                  children: [item],
                };
              }
              currentContainer.children.push(` + `, next);
              // 把当前的节点放到容器内, 然后删除掉j
              children.splice(j, 1);
              // 因为把 j 删除了，所以这里就少了一个元素，那么 j 需要 --
              j--;
            } else {
              break;
            }
          }
        }
      }
    }
  }

}