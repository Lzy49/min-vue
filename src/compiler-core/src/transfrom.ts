import { NODE_TYPE } from "./ast";
import { TO_DISPLAY_STRING, CREATE_ELEMENT_BLOCK } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransfromContext(root, options)

  // 1. 递归 root 找到要修改的节点
  // 2. 为节点 content 增加 'min-vue'
  traverseNode(root, context)
  createRootCodegen(root, context);
  // 将 收集到的 需要增加的 vue 依赖推到 ast 上给 codegen 处理
  root.helpers.push(...context.helpers.values())
}
// 利用插件处理每一点
function traverseNode(node: any, context) {
  // 变化点 抽离为插件
  const exitFns = []
  context.nodeTransforms.forEach(fn => {
    const exitFn = fn(node,context) 
    exitFn && exitFns.push(exitFn)
  })

  // 稳定点 处理子节点
  // 根据不同类型处理节点
  switch (node.type) {
    case NODE_TYPE.INTERPOLATION:
      // 当 template 中有插值行，则需要添加 Vue TO_DISPLAY_STRING 依赖。
      context.helper(TO_DISPLAY_STRING);
      break;
    case NODE_TYPE.ELEMENT:
    case NODE_TYPE.ROOT:
      traverseChildren(node, context);
      break;
  }
  exitFns.forEach(item=>item())
}
function traverseChildren(node: any, context) {
  const children = node.children;
  if (Array.isArray(children)) {
    children.forEach(item => traverseNode(item, context));
  }
}

function createTransfromContext(root, options) {
  // 
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Set(),
    helper: (helper) => {
      context.helpers.add(helper)
    }
  };
  return context
}

function createRootCodegen(root: any, context: { root: any; nodeTransforms: any; }) {
  const child = root.children[0];
  root.codegenNode = child

  if (child.type === NODE_TYPE.ELEMENT && child.codegenNode) {
    const codegenNode = child.codegenNode;
    root.codegenNode = codegenNode;
  } else {
    root.codegenNode = child;
  }
}

