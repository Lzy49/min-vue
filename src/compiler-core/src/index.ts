import { transform } from "../src/transfrom"
import { generate } from "../src/codegen"
import { transformExpression } from '../src/transform/transformExpression'
import { transformText } from '../src/transform/transformText'
import { transformElement } from '../src/transform/transformElement'

import { baseParse } from './parse'
export const baseCompile = (template:string) => {
  const ast = baseParse(template)
  transform(ast, {
    nodeTransforms: [transformExpression, transformText, transformElement,]
  })
  return generate(ast)
}
