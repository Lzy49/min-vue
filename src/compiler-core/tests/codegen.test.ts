import { transform } from "../src/transfrom"
import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transformExpression } from '../src/transform/transformExpression'
import { transformText } from '../src/transform/transformText'
import { transformElement } from '../src/transform/transformElement'
describe('codegen', () => {
  test('happy path', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
  test('interpolation happy path', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformExpression]
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
  test('element happy path', () => {
    const ast = baseParse('<div>hi,{{missage}}</div>')
    transform(ast, {
      nodeTransforms: [transformExpression, transformText, transformElement,]
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})