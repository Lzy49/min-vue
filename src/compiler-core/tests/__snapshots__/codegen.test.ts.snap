// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element happy path 1`] = `"const { toDisplayString: _toDisplayString,createElementBlock: _createElementBlock } = Vue;return function render(_ctx, _cache){return _createElementBlock('div', null, 'hi,' + _toDisplayString(_ctx.missage))}"`;

exports[`codegen happy path 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;

exports[`codegen interpolation happy path 1`] = `"const { toDisplayString: _toDisplayString } = Vue;return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"`;
